js/src/lib.rs

```diff
@@ -51,35 +52,37 @@ impl JsHandle {
     #[allow(non_snake_case)]
     /// Add a permission domain to this `doughnut`
     pub fn addDomain(&mut self, key: &str, value: &[u8]) -> Self {
-        if let Doughnut::V0(mut doughnut) = self.0.clone() {
+        if let Doughnut::V0(ref mut doughnut) = &mut self.0 {
             doughnut.domains.push((key.to_string(), value.to_vec()));
-            return JsHandle(Doughnut::V0(doughnut));
+            return self.clone();
```

Comment:
This looks to be an unnecessary copy which should be left to the users of this method if they need.

Reply:
hmm, this looks to be a required copy (Doughnut doesn't implement Copy trait so that's why `.clone` is used here explicitily). If Doughnut did implement Copy then there would be an implicit copy happening, when returning `self`.
wasm-bindgen can't return by ref, so I guess that is why the copy is required 🤔

Reply:
What I am thinking is when this method is successfully executed (no errors) it modifies the doughnut. Then we should leave it to the places where this method has been called to decide if we need to clone the modified doughnut or not. Basically it seems to be that we are putting two non-cohesive jobs into one method, 1. adding a domain 2. Cloning the doughnut

Reply:
The reason for returning the doughnut is the js API provides chainable method calls e.g.
```js
let doughnut = new Doughnut(..).addDomain().addDomain().signSr25519();
```
The ergonomics > than the cost of copy imo but it's not ideal.
Other alternatives considered:
- Remove method chaining
- Write some JS to wrap these functions and provide chaining (hard to maintain)
- Write specific functions which opt out of the `.clone()` or vice versa (easy to implement if we need it later)

Reply:
I would prefer removing the method chaining. I have noticed that method chaining is not as popular in Rust std as it's in C++. Anyway, I leave this to you to decide and approve this PR.